### PR TITLE
SWAP-1201 CI/CD Improvements

### PR DESCRIPTION
**Pull request by peter-asztalos**
## Description

This PR updates the Github action and Docker configurations to improve the overall speed, size of the images and makes action files consistent.

## Motivation and Context

The previous solution didn't use any caching.

## Fixes

https://jira.esss.lu.se/browse/SWAP-1201

## Changes

- save intermediate layers after each build and use them if there were no changes
- docker ignore updated to provide a better caching experience 
- use multi-stage builds to reduce image size (~290->~160mb)
- package.json updated to use only proper dependencies in each environment
- e2e test related configurations

## Depends on

https://github.com/UserOfficeProject/user-office-backend/pull/52

## Tests included/Docs Updated?

- [ ] I have added tests to cover my changes.
- [ ] All relevant doc has been updated

### CODE REVIEW

**Comment by peter-asztalos** [comment]
@UserOfficeProject/stfc-business-apps 

The dependencies in the package.json changed, many of them are dev dependencies now as we use multi-stage build to reduce the image size.
I think this affects your `Dockerfile.win`.

**Review by jekabs-karklins — APPROVED** [review]

**Review by martin-trajanovski — APPROVED** [review]

File: .github/workflows/test-build.yml [hunk 1]
```diff
@@ -1,60 +1,80 @@
...
-    branches: [ develop, master ]
   pull_request:
-    branches: [ develop, master ]
+    branches: [develop, master]
 
+# A workflow run is made up of one or more jobs that can run sequentially or in parallel
 jobs:
-  build:
-
+  # This workflow contains a single job called "build"
```
Comment (vovsike) [inline .github/workflows/test-build.yml:11]: Is this comment still up to date? It looks like there are two jobs `test` and `build`

File: .github/workflows/test-build.yml [hunk 2]
```diff
@@ -1,60 +1,80 @@
...
+name: Build, Test and Push
 
+# Controls when the action will run. Triggers the workflow on push or pull request
+# events but only for the develop branch
 on:
-  push:
-    branches: [ develop, master ]
   pull_request:
-    branches: [ develop, master ]
+    branches: [develop, master]
```
Comment (vovsike) [inline .github/workflows/test-build.yml:7]: It looks like there are two branches develop and master and the comment above only mentions develop.

File: .github/workflows/build-push.yml [hunk 3]
```diff
@@ -0,0 +1,49 @@
+name: Build && Push
+
+# Controls when the action will run. Triggers the workflow on push or pull request
+# events but only for the develop branch
```
Comment (vovsike) [inline .github/workflows/build-push.yml:4]: It looks like there are two branches develop and master and the comment above only mentions develop.

**Review by vovsike — APPROVED** [review]
Apart from the possible minor comments errors this PR looks good.

**Pull request merged**